lib/spack/llnl/util/tty/log.py

```diff
@@ -89,42 +130,127 @@ def __init__(self, stream):
         """
         self.stream = stream
 
+    def _is_background(self):
+        """True iff calling process is in the background."""
+        return _is_background_tty(self.stream)
+
+    def _get_canon_echo_flags(self):
+        """Get current termios canonical and echo settings."""
+        cfg = termios.tcgetattr(self.stream)
+        return (
+            bool(cfg[3] & termios.ICANON),
+            bool(cfg[3] & termios.ECHO),
+        )
+
+    def _enable_keyboard_input(self):
+        """Disable canonical input and echoing on ``self.stream``."""
+        assert self.old_cfg, "_enable_keyboard_input only works with termios"
+        assert not self._is_background(), \
+            "_enable_keyboard_input can only be called in foreground"
+
+        # "enable" input by disabling canonical mode and echo
+        new_cfg = termios.tcgetattr(self.stream)
+        new_cfg[3] &= ~termios.ICANON
+        new_cfg[3] &= ~termios.ECHO
+
+        # Apply new settings for terminal
+        try:
+            termios.tcsetattr(self.stream, termios.TCSANOW, new_cfg)
+        except termios.error as e:
+            # TODO: does this happen? It was needed at one point, but I
+            # TODO: no longer see it happen.
+            tty.debug("termios error: %s" % e)
+
+    def _restore_input(self):
+        """Restore the original input configuration on ``self.stream``."""
+        assert self.old_cfg, "_restore_input only works with termios"
+
+        # _restore_input Can be called in foreground or background. When called
+        # in the background, tcsetattr triggers SIGTTOU, which we must ignore,
+        # or the process will be stopped.
+        with ignore_signal(signal.SIGTTOU):
+            termios.tcsetattr(self.stream, termios.TCSANOW, self.old_cfg)
+
+    def check_fg(self):
+        if not self.old_cfg:
+            return
+
+        # fix stream settings if we need to
+        flags = self._get_canon_echo_flags()
+        if not self._is_background() and any(flags):
+            self._enable_keyboard_input()
+
+    def _tstp_handler(self, signum, frame):
+        """Handler to restore term settings before caller is backgrounded."""
+        # retore input settings on terminal when backgrounded, so we
+        # don't leave the user's terminal in non-canonical, non-echo mode
+        self._restore_input()
+
+        # Run the default SIGTSTP handler for this process to actually stop it.
+        # We reinstall _tstp_handler on SIGCONT.
+        signal.signal(signal.SIGTSTP, self.old_tstp_handler)
+        os.kill(os.getpid(), signal.SIGTSTP)
+
+    def _cont_handler(self, signum, frame):
+        """Handler to restore or enable term settings on SIGCONT."""
```

Comment:
The docstring diagram for this class is excellent, I think it would be great to refer to it here.

Reply:
This handler no longer exists -- there's just one.